Add iterations parameter to //smooth

sbrl opened the pull request:
Here's try #2!

This PR adds an iterations parameter to `//smooth`, as I've found myself wanting to run many passes at once.

sbrl committed:
Add iterations parameter to //smooth

sbrl committed:
Update README to match

sfan5 commented on README.md:
```diff
@@ -24,7 +24,7 @@ Meaning: Add grass on top and change deeper nodes to stone.
+* `//smooth [<iterations>]`
 
 Smooth terrain in the current selection. When adding nodes, the type of the top-most node in the column is used.
 This uses an exponentially weighted moving average (EWMA) and should be ran 1 or 2 times to produce good results.
```
description should be updated
"should be ran with 1 or 2 iterations"

sfan5 commented on init.lua:
```diff
@@ -197,131 +197,133 @@ local function smooth(pos1, pos2, deadzone)
+	
+	local count = 0
+	for i = 1,iterations do
+		-- read heightmap from data
```
reading/writing the map only needs to be done once

sbrl replied:
Ah, you're right. I didn't notice that! I'll move the heightmap calculation outside the loop.

sbrl committed:
README.md: Improve `//smooth` description

sbrl committed:
Only calculate heightmap once for //smooth

Optimise //smooth by moving the heightmap calculation outside the iteration loop, since the EWMA calculation modifies the heightmap directly

sbrl commented:
Thanks for the review! I've corrected those issues.

sfan5 commented on init.lua:
```diff
@@ -197,7 +197,7 @@ local function smooth(pos1, pos2, deadzone)
 	local c_air = minetest.get_content_id("air")
 	local c_dirt = minetest.get_content_id("default:dirt")
-
+	
```
whitespace

sbrl replied:
Erm you want it removed? Uhhh sure? Not my preference (I think my editor did it automagically), but I can change it here.

sfan5 commented on init.lua:
```diff
@@ -221,107 +221,109 @@ local function smooth(pos1, pos2, deadzone)
-					y = y + 1
+						count = count + 1
+						y = y + 1
+					end
```
writing it back to the map only needs to be done once too...

sbrl replied:
Ah, ok. I'll take a look.

sbrl committed:
Further optimisations: Only write back once

Only write back to the world slice once.

sbrl commented:
Fixed!

sfan5 commented:
I noticed that running just the algorithm multiple times doesn't really work out easily, so I've just implemented this myself.

sbrl commented:
Ah, ok.....? What did you do differently? I'm curious as I've like to improve my own programming skills.

sfan5 commented:
In the current state of this PR, the code will read heights from the heightmap and compute slices in a loop.

There problem here is that later iterations would just read from the heightmap again, completely ignoring the previous results. The second problem is, the final heights are calculated inside the loop that "writes" heights back to the map, so that needs to be done multiple times too.

What I've done is to refactor the algorithm into a separate method that does both: compute slices and calculate the new heightmap.
Then I can easily call the method in a loop, each time reusing and amending the previous results.

sbrl commented:
Ahhh, I see. Thanks for the explanation :smiley_cat:
